substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/PredicatedConfigurationSupport.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2021, 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.core;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
+import org.graalvm.nativeimage.hosted.Feature;
+import org.graalvm.nativeimage.impl.ConfigurationPredicate;
+
+@Platforms(Platform.HOSTED_ONLY.class)
+public abstract class PredicatedConfigurationSupport {
+    /* Keep track of classes already processed for reflection. */
+    private final Map<String, List<Runnable>> reachabilityHandlers = new ConcurrentHashMap<>();
+
+    public void registerPredicatedConfiguration(Feature.BeforeAnalysisAccess b) {
+        for (Map.Entry<String, List<Runnable>> reachabilityEntry : reachabilityHandlers.entrySet()) {
+            TypeResult<Class<?>> typeResult = findClass(b, reachabilityEntry.getKey());
+            b.registerReachabilityHandler(access -> reachabilityEntry.getValue().forEach(Runnable::run), typeResult.get());
+        }
+        reachabilityHandlers.clear();
+    }
+
+    protected abstract TypeResult<Class<?>> findClass(Feature.BeforeAnalysisAccess b, String className);
```

Comment:
I'm not quite sure why `findClass` needs to be `protected abstract` because currently both implementations do the same.

Reply:
Because we don't see the `BeforeAnalysisAccessImpl` from `core`. If we want to reduce this duplication, we can make a common super-class with this one method?

Reply:
Why does this class have to be in `core`? Both subclasses can access `hosted` classes and it looks like this class could be in `hosted`, too.

Reply:
Fixed.